Patch rare refresh exception

13steinj opened the pull request:
Resolves a rare refresh exception upon refreshing extremely unique comments, as well as stops the InvalidComment exception from occuring when a `Comment`'s `replies` property is called and it's replies weren't previously fetched.

~~I still want to add 3 tests (assertRaises on get_info and assertWarningsRegexp to ensure that there are keys that can not be updated, such as `was_comment`, in both the deleted and the removed case), but I thought I'd make a PR now to discuss caveats.~~

This patch *heavily* relies on `get_info`, but `get_info` just couldn't get the job done properly in some cases because of the fact that there was no way to burst the cache via the `uniq` parameter, since there were no args or keywordarguments that could be passed.  Also, `get_info` had a bug regarding it's use. The `limit` paramter actually *can* be used when using thing_ids, and as weird as it sounds, it does make sense (assume that the user's default preference is 25 items per listing, and they use `get_info` with 37 valid ids, to actually get all 37 they'd need to pass the limit parameter  as None or 37). With the above and the fact that it was inconsistent with *all* other casual functions that retrieve from reddit's Listing json kind, I altered the functionality to return a `get_content` generator instead, allow for args and kwargs to be passed on to it, and stop the faulty restriction with the limit param, and altered the tests to fit (but now looking at it I could have alterred them better, so I'll do that).

Also, I am not sure if the warnings would be pushed through a second time if they are called upon a second time (I don't know how the warnings module works to that extent).

I had to update the betamax version used for tests, because sometime between 0.4.2 and 0.5.1 they had a faulty release that broke all tests. I know I could have downgraded to 0.4.2, but I thought upgrading would have been better.

Edit: I've added the tests that I wanted to add.

bboe commented on praw/__init__.py:
```diff
@@ -887,43 +887,35 @@ def get_front_page(self, *args, **kwargs):
-        :returns: When a single ``thing_id`` is provided, return the
-            corresponding thing object, or ``None`` if not found. When a list
-            of ``thing_id``s or a ``url`` is provided return a list of thing
-            objects (up to ``limit``). ``None`` is returned if any one of the
```
Should we mention that any one invalid ID can cause the function to yield nothing?

13steinj replied:
@bboe IIRC, as long as there is a valid id it still works, it just yields nothing if *all* ids are invalid.

bboe replied:
Interesting. That must have changed at some point.

bboe commented on praw/__init__.py:
```diff
@@ -887,43 +887,35 @@ def get_front_page(self, *args, **kwargs):
-        else:
-            return None
+
+        if 'params' in kwargs:
```
I don't think this is necessary.

bboe commented on praw/objects.py:
```diff
@@ -441,7 +441,31 @@ def refresh(self):
             sub = Submission.from_url(self.reddit_session, self.permalink,
                                       params={'uniq': unique})
-            other = sub.comments[0]
+            try:
```
Can we use an `if / else` here instead of exception handling?

bboe commented on praw/__init__.py:
```diff
@@ -887,43 +887,35 @@ def get_front_page(self, *args, **kwargs):
+
+        if 'params' in kwargs:
+            params.update(kwargs['params'])
+            kwargs.pop('params')
```
This line isn't necessary due to line 916.

bboe commented on praw/objects.py:
```diff
@@ -441,7 +441,31 @@ def refresh(self):
+                newkeys = set(other.__dict__.keys())
+                keydiff = ", ".join(oldkeys - newkeys)
+                if keydiff:
+                    msg += "\n" + "Could not retrieve:\n" + keydiff
```
Can you use the format string here (e.g., `'\nCould not... : {0}'.format(keydiff)`)?

bboe commented on tests/helper.py:
```diff
@@ -54,6 +56,8 @@ def configure(self):
         self.redirect_uri = 'https://127.0.0.1:65010/authorize_callback'
 
         self.comment_url = self.url('/r/redditdev/comments/dtg4j/')
+        self.reply_warnings_url = self.url(('/r/reddit_api_test/'
```
No need for the inner parens in this case.

bboe commented on tests/test_message.py:
```diff
@@ -15,7 +15,8 @@ def betamax_init(self):
         comment_reply = next(self.r.get_comment_replies(limit=1))
-        comment_parent = self.r.get_info(thing_id=comment_reply.parent_id)
+        info_id = comment_reply.parent_id
+        comment_parent = next(self.r.get_info(thing_id=info_id))
```
It'd be nice if when `thing_id` only contains a single item, that `get_info` still returns that one item. Then we don't break this part of the behavior.

bboe commented:
Great PR. I like this addition. My only concern is that the conversion of `get_info` to return a generator will make it such that we have to postpone merging this until PRAW4 (or create a separate branch). If we do away with that part of the change, then we should be able to merge this in sooner.

Doesn't `get_info` have a maximum limit of number of items one can provide to it? Is there really ever a reason that it would require more than 1 request to get all the infos for the items? If there isn't then, I see no reason to use the `get_content` function at all. Instead, we can ensure the limit is sufficient for the request size.

13steinj commented:
@bboe There is a maximum amount of ids you can pass (100  as shown [here](https://github.com/reddit/reddit/blob/master/r2/r2/controllers/api.py#L170)), but the number of *items returned* is not limited anywhere that I can tell (except maybe the 1000 performance based limit, which is the same limit for /new and /top and such). While purely theoretical, there's a chance that /api/info?url=https://niceme.me could have 32 items, but it can also have 789 items.

Also given that I can't find a decent way to ensure that if there's only one item to return only one item. Even if we disregard url lookups for that, I've no clue on how to ensure that there's only one real argument (ie, no accidental commas because checking and then filtering for those are a pain, and checking on a list would only be made harder, since theoretically the list passed can be `['fakeid1,','fakeid2','fakeid3,,', 'fakeid4,fakeid5']` (that list *works*, but wouldn't be easy to distinguish to return a single item)

I'm all for such an option that would actually raise the errors, but I don't know a good way to implement it.

---

On a semi separate note, with all these deprecative changes propping up slowly, it may be good to invest in a decorator `@legacy` for the sake of a name, and if the keyword argument `legacy=True` is in any function that is to have a change, use `function_name-legacy` instead that is located in a separate file (?`legacy.py`), which would use the legacy behavior, and, if it isn't passed, trigger a deprecation warning? It technically solves the problem of "a change is good / wanted / needed, but breaks other behavior".

bboe commented:
@13steinj can you confirm that the info for a URL can have more than 100 items returned thus making a need for a generator. If that's in-fact the case, maybe we should split the client-level function for the two.

Regarding returning a single item, the only case there should be more than one item for fetching by ID is if a non-string argument is passed in. The behavior has always assumed a string argument represents a single ID, and that each string in a list also represents a single id. Thus if it's passed an ID as a string, it gets the element immediately, otherwise it gets a list that should contain the same number of elements as their list.

13steinj commented:
@bboe now that I'm able to get back into the swing of things, I've rebased and updated to keep behavior the same.

I just still didn't add an option to replace those warnings with errors, since I don't know where that should be set (from the config file?)

bboe commented on tests/cassettes/test_raise_invalidcomment_oauth.json:
```diff
@@ -1,11 +1,11 @@
 {
   "http_interactions": [
     {
```
Why is this test updated?

bboe commented on tests/test_comments.py:
```diff
@@ -180,8 +180,103 @@ class OAuthCommentTest(OAuthPRAWTest):
     def test_raise_invalidcomment_oauth(self):
         fullname = '{0}_{1}'.format(self.r.config.by_object[Comment],
                                     self.comment_deleted_id)
-        self.r.refresh_access_information(self.refresh_token['submit'])
```
Does this need to be reordered?

13steinj replied:
The cassette did not need to be changed however this needs to be reordered as it seems that previously a 'submit' access token was allowed on /api/info but now it requires the read scope and that alone.

bboe commented:
@13steinj looks like the PR needs to be rebased again. I had one question about why a test case and it's cassette were updated.

I'm kind of apathetic about this change as much of the work might not be necessary with PRAW4. And it seems rare enough that changing the return type (if that's still correct) is not something that should be done in a minor version. How often have you encountered this exception?

13steinj commented:
@bboe I'll look at the test case when I'm at a pc, changing that was probably by accident.

I've reverted the change to the return type, so the behavior of `get_info` isn't altered as per your previous mention of that issue.

Also in regards to PRAW4, I can take a look and port the necessary fix in a separate PR if the issue exists there.

bboe commented:
Awesome. Thanks for following through with it.

13steinj commented:
@bboe I've looked at PRAW4, and the issue doesn't exist because as of right now there isn't a `refresh` method added by some form of `RefreshMixin` (is the lack of such intentional?), however a fix may need to be considered when [this TODO is](https://github.com/praw-dev/praw/blob/praw4/praw/models/reddit/comment.py#L53), as comments in an inbox also share that discrepancy.

bboe commented:
Refresh was removed to keep things simple. I think it will make sense to add back in later if it can be done simply.
